### ISSUE TITLE
Current Status 7/30/15

### ISSUE THREAD

**Issue by mfilosa** [mfilosa/Monitor-Docs#3]
SCMon Team @beetle-ibm 

Currently parsing unicorn.log and audit.log 

unicorn.log for web traffic
audit.log for pushes and pulls.

**Comment by beetle-ibm** [comment]
in audit.log I see entries that are unlike normal entries that look like this
 hook_ids="3,5" hook_names="web,web"
what are these?

**Comment by beetle-ibm** [comment]
FYI I think your example of a git push in the audit log is wrong.  It should be a git-receive-pack, not a git-upload-pack.

**Comment by beetle-ibm** [comment]
It appears there is no year in the audit.log.  Can this be added?

**Comment by beetle-ibm** [comment]
on pull I see these...
"uploaded_bytes":453
"uploaded_bytes_last":1439245473 
What do these mean? 

Also I see status fields like this
"status":"create_pack_file"
"status":"post_receive_hook"

What do these mean?  Can we tell from these if a push/pull was successful?

**Comment by beetle-ibm** [comment]
In the audit.log I see fields user_login, user_id, and user... user_login only seems to be there when its a push or a pull.  I see user_id and user some other places... is there some important distinction between these?  (namely user_login, and user... I see that user_id is numeric)

**Comment by beetle-ibm** [comment]
will the repo part of the link always be a namespace/repo name? Or is there a possibility for further nesting/tree structures?  

For example here is the link in the log from downloading a snapshot. 
http://scmon-github.rtp.raleigh.ibm.com/**beetle-us/scmon-test**/archive/master.zip

I would identify the repo as "beetle-us/scmon-test"
Will it ALWAYS be this way?  

Or is if ever possible for a repo to exist at the root?  e.g.
http://scmon-github.rtp.raleigh.ibm.com/**scmon-test**/archive/master.zip

Or for a repo to be further down a namespace/tree structure
http://scmon-github.rtp.raleigh.ibm.com/**team-area/beetle-us/scmon-test**/archive/master.zip

**Comment by ddumontatibm** [comment]
@mfilosa Any updates on these questions?

**Comment by mfilosa** [comment]
@ddumontatibm we have an issue open internally to try to answer these. This requires some testing of different scenarios and seeing how the logs behave. If you have found any answers to the above questions could you please reply so that I can spend out time appropriately.

**Comment by mfilosa** [comment]
> in audit.log I see entries that are unlike normal entries that look like this
hook_ids="3,5" hook_names="web,web"
what are these?

These are hooks that are triggered. Here is the relevant API you should be able to query the details by Id if you need to: https://developer.github.com/v3/repos/hooks/

**Comment by mfilosa** [comment]
That's -- odd. the timestamp should have a year. I don't have access to an example, can you paste some?

> In the audit.log I see fields user_login, user_id, and user... user_login only seems to be there when its a push or a pull. I see user_id and user some other places... is there some important distinction between these? (namely user_login, and user... I see that user_id is numeric)

We don't have normalized log output from all the various services that are shipping logs to `audit.log`. I'm going to hazard a guess that some of these are configured to use `user_login` and some to use `user_id`

> What do these mean? Can we tell from these if a push/pull was successful?

We could probably determine from the amount of bytes uploaded if the push/pull was successful. Would it be more valuable to find out what does a failed push/pull look like in the log? We can try to fail a push / pul and see if the log reports something differently?

**Comment by beetle-ibm** [comment]
I am not very familiar with hooks... basically you are saying that those log entries mean that a hook was run after someone pushed a commit or something like that?  So basically it does not represent a user action but rather an automated reaction to a user action... is that right?  

If you could easily produce some failures that would be great to see what they look like.  When you speak of basing it on bytes are you implying that a failure would always be bytes=0?  

Speaking of bytes any update on which bytes parameter is the one we should be looking at?  uploaded_bytes or uploaded_bytes_last?  


FYI example of log output with no year
<code>Aug  5 06:33:41 git-swg-usma-ibm-com github_audit[48645]: { "pid":48642, "ppid":48641, "program":"upload-pack", "git_dir":"/data/user/repositories/c/nw/c4/ca/42/1/1.git", "cmdline":"git-upload-pack --strict --timeout=0 --stateless-rpc .", "repo_name":"IBM-Verse/sequoia", "hostname":"git-swg-usma-ibm-com", "remote_addr":"127.0.0.1", "remote_port":"17445", "babeld":"e8e156d", "real_ip":"9.115.192.89", "transaction_id":"f8593d606e43ccc8be015545785ce4b0733ffd1875016c30", "frontend":"git-swg-usma-ibm-com", "frontend_pid":25845, "frontend_ppid":5467, "repo_id":1, "repo_public":false, "repo_guest_branches_enabled":false, "repo_deny_default_branch":false, "repo_default_branch":"master", "repo_config":"{\"disable_user_org_creation\":\"true\",\"git.reject_force_push\":\"default\",\"git-lfs\":\"true\"}", "user_id":78, "user_login":"wzhaoch", "user_operator_mode":false, "gitauth_version":"9b37fa43", "pgroup":"48642", "status":"create_pack_file", "features":" include-tag ofs-delta multi_ack_detailed no-done thin-pack side-band-64k", "cloning":false, "uploaded_bytes":42068, "uploaded_bytes_rate":0, "uploaded_bytes_last":1438756421 }</code>

**Comment by mfilosa** [comment]
@beetle-ibm 

> If you could easily produce some failures that would be great to see what they look like. When you speak of basing it on bytes are you implying that a failure would always be bytes=0?

I am thinking that the comparing the bytes in the log vs the expected bytes would help determine a failure. However I have some time today to experiment I will post what I find here.

**Comment by beetle-ibm** [comment]
How would we know expected bytes?  Most pulls will be pulling updates, and we won't know how many bytes of updates to expect.  It would be best to just have a flag to tell us if an action succeeded for failed.  In any case though, the whole status thing is less important that knowing which bytes variable is the one we should be looking at for response size.  Thanks!

**Comment by mfilosa** [comment]
> I am not very familiar with hooks... basically you are saying that those log entries mean that a hook was run after someone pushed a commit or something like that? So basically it does not represent a user action but rather an automated reaction to a user action... is that right?

That is correct a webhook is essentially a way to integrate GitHub with other systems. For example after a commit is made a webhook may get fired that will post the commit message and information to a third party system e.g. IBM RTC Defect tracker or Jira or trigger. 

The message is a HTTP(s) post with a payload of JSON with the receiving system then processes and performs an action. That system can the call back to GitHub Enterprise via the [developer API] (https://developer.github.com/v3/) 

A example of this would be a commit to a branch triggering a CI job to execute in Jenkins or Travis CI that then reports back the status of the build to GitHub via the status API provided by GitHub.

For a test you can easily create a webhook on a repository and trigger is by performing an action you have configured to fire the hook. 

To set up a repository webhook on GitHub, head over to the Settings page of your repository, and click on Webhooks & services. After that, click on Add webhook.

Alternatively, you can choose to build and manage a webhook through the Webhooks API.

Further information can be found here: https://help.github.com/enterprise/2.3/user/articles/about-webhooks/
